shared/runtime/pyexec.c

```diff
@@ -143,11 +147,15 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
             mp_hal_stdout_tx_strn("\x04", 1);
         }
 
-        // check for SystemExit
-        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) {
+        #ifdef MICROPY_ENABLE_VM_ABORT
+        if (nlr.ret_val == NULL) { // abort
+            ret = 255;
```

Comment:
Probably want to define a macro for this value and document it.

Unhanded exceptions still return ret = 0 though, so why is this different?

Reply:
And this should probably be followed by a `goto` that jumps to a label just before `MICROPY_BOARD_AFTER_PYTHON_EXEC()` at the end of the function so that no other MicroPython runtime stuff runs.

Reply:
> Probably want to define a macro for this value and document it.

Ok. Where is a good place to define it?

Reply:
> Unhanded exceptions still return ret = 0 though, so why is this different?

Is this on purpose? I consider it a bug. In my port/usecase I want the exceptions to throw a non-success exit code and I was planning opening an additional PR later. As well as exposing the exit code from system exit, so `os.exit(1)` in python actually exits with the exit code according to the argument.

Reply:
> And this should probably be followed by a goto that jumps to a label just before MICROPY_BOARD_AFTER_PYTHON_EXEC() at the end of the function so that no other MicroPython runtime stuff runs.

Is it wrong to run the repl stuff? And the `mp_hal_stdout_tx_strn` function. Those are the only two things that are potentially run

Reply:
> > Probably want to define a macro for this value and document it.
> 
> Ok. Where is a good place to define it?

At the moment, I implemented it the same way as `extern int pyexec_system_exit;` is set up. Do you think this is appropriate?

Reply:
> Where is a good place to define it?

Maybe in `mpconfig.h` right after `MICROPY_ENABLE_VM_ABORT`? Then users could override the default value if needed.

>  I implemented it the same way as extern int pyexec_system_exit; Do you think this is appropriate?

Only if there is an a actual need to know different possible reasons for the abort. But if we do need that info, then it seems like the way to do it.

> Is this on purpose?

Not sure, but if you could explain more about your use case that could help the maintainers decide if this is something we should change or not.


> Is it wrong to run the repl stuff?

Again, I think it depends on the use case. Perhaps this is why Damien left this code out when he implemented the feature. There isn't are obvious "right way" to me.

Reply:
My use case is this:

I am building a device which is programmable by micro python by the end users. The device/firmware provides an additional "OS" layer - display/UI/menu/settings/file system/... and API for the user program. When the user program is being executed, it is rendered on the display, sending stdout to the display and other info. When the program exits, I want to be able to show a screen "Finished successfully" or "Crashed" (triggered by an unhandled exception or os.exit(non-zero) or "Killed/aborted". The abort is triggered from the device itself/button combination. I need the abort to be instant (as perceived by the user). Which means, there is some time to finish some instructions, flush std out, ...; However no more python instructions should be run (so the program wont be able to "prevent" being killed)

Reply:
For this, it makes sense to me to return exit codes on the exit. Not just always zero. I am not sure if there is a use case, where you would want to have always zero exit code. If the host fw/vm doesn't care about it, it can just ignore it. Right?

Am I missing something?

Reply:
> I am building a device which is programmable by micro python by the end users.

We did something similar with Pybricks. What we ended up doing was just raise `SystemExit` to stop user programs. We only use the abort if we are powering off the device. (If a user manages to somehow always catch the `SystemExit`, they will have to power off the device, but this doesn't really seem to happen that we've notified).

Here is the source if you are interested: https://github.com/pybricks/pybricks-micropython/blob/79a0e010ac36fb64593d5ef5b213b64f1a03985b/bricks/_common/micropython.c#L229

> If the host fw/vm doesn't care about it, it can just ignore it. Right?

Makes sense to me.

Reply:
I see. Should one be preferred over the other? Or is this approach OK as well?
Is there anything blocking this merge at the moment?

Reply:
I think we will have to wait for @dpgeorge to chime in here (sometimes it can take a while before he has time).